### ISSUE TITLE
Non-deterministic output with the streaming model

### ISSUE THREAD

**Issue by reuben** [mozilla/DeepSpeech#1563]
Example:

```
./deepspeech --model ~/Development/DeepSpeech/1300_export/output_graph.pbmm --alphabet ~/Development/DeepSpeech/data/alphabet.txt --lm ~/Development/DeepSpeech/data/lm/lm.binary --trie ~/Development/DeepSpeech/data/lm/trie --audio audio/repeat
TensorFlow: v1.6.0-18-g50214731ea
DeepSpeech: unknown
2018-09-19 13:54:04.607023: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Running on directory audio/repeat
> audio/repeat/16.wav
why should one hall on the dway
> audio/repeat/17.wav
why should one hall on the dway
> audio/repeat/15.wav
why should one hall on the dway
> audio/repeat/29.wav
why should one hall on the dway
> audio/repeat/28.wav
why should one hall on the dway
> audio/repeat/14.wav
why should one hall on the dway
> audio/repeat/10.wav
why should one hall on the dway
> audio/repeat/11.wav
why should one hall on the dway
> audio/repeat/13.wav
why should one hall on the dway
> audio/repeat/12.wav
why should one hall to on the way'
> audio/repeat/9.wav
why should one hall on the dway
> audio/repeat/8.wav
why should one hall on the dway
> audio/repeat/6.wav
why should one hall on the dway
> audio/repeat/7.wav
why should one hall on the dway
> audio/repeat/5.wav
why should one hall on the dway
> audio/repeat/4.wav
why should one halt on the way'
> audio/repeat/1.wav
why should one hall on the dway
> audio/repeat/3.wav
why should one hall on the dway
> audio/repeat/2.wav
why should one hall on the way'
> audio/repeat/23.wav
why should one hall on the dway
> audio/repeat/22.wav
why should one hall on the dway
> audio/repeat/20.wav
why should one hall on the dway
> audio/repeat/21.wav
why should one hall on the dway
> audio/repeat/19.wav
why should one hall on the dway
> audio/repeat/25.wav
why should one halt on the way'
> audio/repeat/24.wav
why should one hall on the dway
> audio/repeat/30.wav
why should one halt on the way'
> audio/repeat/18.wav
why should one hall on the dway
> audio/repeat/26.wav
why should one hall to on the way'
> audio/repeat/27.wav
why should one hall on the dway
```

**Comment by wefgy** [comment]
./deepspeech --model ~/Development/DeepSpeech/1300_export/output_graph.pbmm --alphabet ~/Development/DeepSpeech/data/alphabet.txt --lm ~/Development/DeepSpeech/data/lm/lm.binary --trie ~/Development/DeepSpeech/data/lm/trie --audio audio/repeat
TensorFlow: v1.6.0-18-g50214731ea
DeepSpeech: unknown
2018-09-19 13:54:04.607023: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Running on directory audio/repeat
> audio/repeat/16.wav
why should one hall on the dway
> audio/repeat/17.wav
why should one hall on the dway
> audio/repeat/15.wav
why should one hall on the dway
> audio/repeat/29.wav
why should one hall on the dway
> audio/repeat/28.wav
why should one hall on the dway
> audio/repeat/14.wav
why should one hall on the dway
> audio/repeat/10.wav
why should one hall on the dway
> audio/repeat/11.wav
why should one hall on the dway
> audio/repeat/13.wav
why should one hall on the dway
> audio/repeat/12.wav
why should one hall to on the way'
> audio/repeat/9.wav
why should one hall on the dway
> audio/repeat/8.wav
why should one hall on the dway
> audio/repeat/6.wav
why should one hall on the dway
> audio/repeat/7.wav
why should one hall on the dway
> audio/repeat/5.wav
why should one hall on the dway
> audio/repeat/4.wav
why should one halt on the way'
> audio/repeat/1.wav
why should one hall on the dway
> audio/repeat/3.wav
why should one hall on the dway
> audio/repeat/2.wav
why should one hall on the way'
> audio/repeat/23.wav
why should one hall on the dway
> audio/repeat/22.wav
why should one hall on the dway
> audio/repeat/20.wav
why should one hall on the dway
> audio/repeat/21.wav
why should one hall on the dway
> audio/repeat/19.wav
why should one hall on the dway
> audio/repeat/25.wav
why should one halt on the way'
> audio/repeat/24.wav
why should one hall on the dway
> audio/repeat/30.wav
why should one halt on the way'
> audio/repeat/18.wav
why should one hall on the dway
> audio/repeat/26.wav
why should one hall to on the way'
> audio/repeat/7.htc
why should one hall on the dway

**Comment by kdavis-mozilla** [comment]
Another data point, but this time in the opposite direction.

Compiled v0.2.0 on Ubuntu 16.04, Deep Learning base image from AWS, and ran the native client 80 times over the 8455-210777-0068.wav example audio, each time obtaining the same result.

So maybe it's the wrapping of the native code in Python, Node, or Rust? Or maybe it's some mismatch between the target platform and the platform the code was compiled for?

**Comment by reuben** [comment]
I've disabled using the LM decoder in the prod tests as we were getting intermittent failures, we should go back to using the LM there when this is fixed.

**Comment by lissyx** [comment]
> I've disabled using the LM decoder in the prod tests as we were getting intermittent failures, we should go back to using the LM there when this is fixed.

FTR, there is still some intermittence even after disabling the LM. It looks less frequent, though.

**Comment by lissyx** [comment]
Update: looking at retriggers failures on https://tools.taskcluster.net/groups/UgiKCb1TSYuMCy0waX5o-g, it would looks like only NodeJS and Python prod are impacted.

C++, NodeJS and Python prod tests are all using the LM through the same `run_prod_inference_tests` function. Though, as much as I can tell, in **all** of the failures reported in the above link, it is when we test the inference on the 44kHz to 16kHz downsampling.

C++ links directly to `libsox` while NodeJS and Python more or less relies on `stdout` of `sox` binary. Maybe there is a slight variation and it triggers intermittence. At that point, I'm starting to think that the TaskCluster intermittence is unrelated to the one reported here because of the LM ?

**Comment by lissyx** [comment]
@kdavis-mozilla I have been able to reproduce locally with v0.2.0 C++ binaries and the use of LM, so if there's anything, it's unrelated to Python, Node or Rust and it's lower-level.

**Comment by kdavis-mozilla** [comment]
@lissyx As an addendum to my comment. I ran the native client 80 times. But each time I restarted the client. I did not pass a directory full of wav files to the client.

@reuben on the other hand passed a directory to the client so his client was never restarted. So it seems there is some language model state that is not properly "re-set" when multiple wav file are decoded in a single client run.

**Comment by lissyx** [comment]
> @lissyx As an addendum to my comment. I ran the native client 80 times. But each time I restarted the client. I did not pass a directory full of wav files to the client.

Yep, that could be consistent, I did reproduce on a full directory pass, and this was after the 95th iteration.

i'll go and play with RR now that the TaskCluster issue has been addressed and we know it's unrelated.

**Comment by lissyx** [comment]
[deepspeech.valgrind.log](https://github.com/mozilla/DeepSpeech/files/2423393/deepspeech.valgrind.log)

**Comment by lissyx** [comment]
I'm still reproducing even with the patches from #1600

**Comment by lissyx** [comment]
looks like #1602 will do it :)

**Comment by lissyx** [comment]
Fixed by #1603

**Comment by est31** [comment]
Thanks for the fix!
